usr/share/rear/lib/global-functions.sh

```diff
@@ -14,6 +14,38 @@ function read_and_strip_file () {
     sed -e '/^[[:space:]]/d;/^$/d;/^#/d' "$filename"
 }
 
+# Output lines in STDIN or in a file without subsequent duplicate lines
+# i.e. for each line that was seen (and output) do not output subsequent duplicates of that line.
+# This keeps the ordering of the lines so the input
+#   one
+#   two
+#   one
+#   three
+#   two
+#   one
+# gets output as
+#   one
+#   two
+#   three
+# To remove duplicate lines and keep the ordering one could use ... | cat -n | sort -uk2 | sort -nk1 | cut -f2-
+# cf. https://stackoverflow.com/questions/11532157/remove-duplicate-lines-without-sorting/11532197#11532197
+# that also explains an awk command that prints each line provided the line was not seen before.
+# The awk variable $0 holds an entire line and square brackets is associative array access in awk.
+# For each line the node of the associative array 'seen' is incremented and the line is printed
+# if the content of that node was not '!' previously set (i.e. if the line was not previously seen)
+# cf. https://www.thegeekstuff.com/2010/03/awk-arrays-explained-with-5-practical-examples/
+function unique_unsorted () {
+    local filename="$1"
+    if test "$filename" ; then
+        test -r "$filename" && awk '!seen[$0]++' "$filename" || return 1
+    else
+        # Timeout after 5 seconds when there is nothing at STDIN or when STDIN does not get closed
+        # to avoid that 'awk' waits endlessly for (further) input which makes ReaR behave as if it hung up
+        # cf. https://github.com/rear/rear/wiki/Coding-Style#beware-of-the-emptiness
+        timeout 5 awk '!seen[$0]++'
```

Comment:
While this works as intended it might be better to check for unintented scenarios with `tty` and abort if stdin is a tty?
![image](https://github.com/rear/rear/assets/101384/cf8561d6-b19c-4d0f-8d92-ea849b0d9dc1)

I'm worried about difficulties identifying bugs where ReaR just hangs 5 secs which could be hard to distinguish from ReaR actually doing somthing.

Reply:
What about a timeout of only one second?
This should be more than enough time for our use cases.

I copied the "timeout 5" from
init/default/100_check_stale_nfs_mounts.sh
where we Error out when things don't behave as they should.

I think I can add some explicit exit code handling
when it times out because "man timeout" tells
```
If the command times out,
and --preserve-status is not set,
then exit with status 124.
Otherwise, exit with the status of COMMAND.
```
so I think it is safe to assume that exit code 124
happens only when it times out because "man awk" tells
```
EXIT STATUS

If the exit statement is used with a value,
then gawk exits with the numeric value given to it.

Otherwise, if there were no problems during execution,
gawk exits with the value of the C constant EXIT_SUCCESS.
This is usually zero.

If an error occurs, gawk exits with the value of the
C constant EXIT_FAILURE. This is usually one.

If gawk exits because of a fatal error, the exit status is 2.
On non-POSIX systems, this value may be mapped to EXIT_FAILURE.
```
so in the awk use case here (no exit statement)
the awk exit codes can be only 0, 1, or 2.

Reply:
Checking with 'tty' won't help against
too slow "dripping" input via a pipe from a program
or when the producer program in a pipe hung up
(so it does not close its stdout).

Reply:
Honestly:
Sine I learned about 'timeout' I like to experiment
with it how far that actually helps in practice.
In particular because 'timeout' could be a generic method
to deal with all kind of "unexpected hangup" issues, cf.
https://github.com/rear/rear/wiki/Coding-Style#it-should-be-possible-to-run-rear-unattended

Reply:
I understand why `timeout` seems like a good solution to many problems, however I'd like to caution against using it for *internal* stuff because it introduces non-deterministic behavior and that can create hard-to-analyze bugs.

For activities that interact with *external* components, e.g. network access, however `timeout` is a good idea to set an upper boundary for stuff that might otherwise block forever.

Reading and sorting file seems for me to be an "internal" problem so that I wouldn't use a timeout solution for that.

Maybe another example: Every use of `cat` or `read` in ReaR carries the same risk that it would silently wait for input from STDIN, and we don't have the habit of protecting against such cases. Therefore I would also suggest to simply ignore that scenario with your sorting function and only make sure that it handles both reading from STDIN and from a file, similar to how `cat` behaves.

Reply:
I agree.

For now I leave out checking for possible problems
when input is stdin.

Perhaps a specific check if stdin is a tty
is really the best compromise by catching
only the one most often happeneing misuse case
and not overengineering things with an attempt
to find a generic solution to catch all misuse cases.
But I will add such a specific check later after
thinking a bit more about it.

For now I merge it as is because its current state
should be good enough - at least for now.